### PR TITLE
replay on the old block without auto gas

### PR DESCRIPTION
**Pull request by ypenghui6**
This change is [Reviewable](https://reviewable.io/reviews/cocos-bcx/cocos-mainnet/76)

### CODE REVIEW

File: libraries/chain/include/graphene/chain/hardfork.hpp [hunk 1]
```diff
@@ -21,3 +21,7 @@
 #ifndef UNSUCCESSFUL_CANDIDATE_DIFFPOINT
 #define UNSUCCESSFUL_CANDIDATE_DIFFPOINT 1765771
 #endif
+// AUTO GAS
+#ifndef UNAUTO_GAS
+#define UNAUTO_GAS (fc::time_point_sec( 1585826672 ))
+#endif
```
Comment (wanliqun) [inline libraries/chain/include/graphene/chain/hardfork.hpp]: Why use such name as UNAUTO_GAS?  Would be much better to use a more explicit name.

File: libraries/chain/hardfork.d/20200320.hf [hunk 2]
```diff
@@ -0,0 +1,4 @@
+/* 于20200402自动回收gas，在此区块之前的按需手动回收；在此之后的，自动回收gas*/
+#ifndef UNAUTO_GAS
+#define UNAUTO_GAS (fc::time_point_sec( 1585826672 ))
+#endif
```
Comment (wanliqun) [inline libraries/chain/hardfork.d/20200320.hf]: I observed this macro had also been defined in libraries/chain/include/graphene/chain/hardfork.hpp. Why do we need to define this in two different places?

File: libraries/chain/db_block.cpp [hunk 3]
```diff
@@ -822,20 +822,23 @@ processed_transaction database::_apply_transaction(const signed_transaction &trx
...
+          account_id_type op_from;
+          if( call_contract_condition ){
+            op_from = op.get<call_contract_function_operation>().caller;
+          }
+          if( transfer_condition ){
+            op_from = op.get<transfer_operation>().from;
+          }
+          if(last_from != op_from){
+            auto_gas(eval_state, op_from);
+            last_from = op_from;
```
Comment (wanliqun) [inline libraries/chain/db_block.cpp:840]: How about code this like this to keep the code more robust:
if(auto_gas(eval_state, op_from)) {
    last_from = op_from;
}

File: libraries/chain/db_block.cpp [hunk 4]
```diff
@@ -867,37 +870,45 @@ processed_transaction database::_apply_transaction(const signed_transaction &trx
   FC_CAPTURE_AND_RETHROW((trx))
 }
 
-bool database::auto_gas(transaction_evaluation_state &eval_state, account_id_type from){
+void database::auto_gas(transaction_evaluation_state &eval_state, account_id_type from){
```
Comment (wanliqun) [inline libraries/chain/db_block.cpp:873]: I think bool return is ok for context like this

**Review by ypenghui6 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cocos-bcx/cocos-mainnet/76)* status: 0 of 4 files reviewed, 4 unresolved discussions (waiting on @wanliqun)

___
*[libraries/chain/db_block.cpp, line 840 at r1](https://reviewable.io/reviews/cocos-bcx/cocos-mainnet/76#-M2qs2o6a4oV1qrC4u5f-r1-840:-M2rLTCnEh7TKnmx_zCG:b-896fix) ([raw file](https://github.com/cocos-bcx/cocos-mainnet/blob/8335277d80aa74ffbf4755a4b8f1ee9047d7cc5f/libraries/chain/db_block.cpp#L840)):*
<details><summary><i>Previously, wanliqun (wanliqun) wrote…</i></summary><blockquote>

How about code this like this to keep the code more robust:
if(auto_gas(eval_state, op_from)) {
    last_from = op_from;
}
          
</blockquote></details>

Done.
___
*[libraries/chain/db_block.cpp, line 873 at r1](https://reviewable.io/reviews/cocos-bcx/cocos-mainnet/76#-M2qs2o6a4oV1qrC4u5f-r1-873:-M2rLTgH3B6wyOyzQmPM:b-896fix) ([raw file](https://github.com/cocos-bcx/cocos-mainnet/blob/8335277d80aa74ffbf4755a4b8f1ee9047d7cc5f/libraries/chain/db_block.cpp#L873)):*
<details><summary><i>Previously, wanliqun (wanliqun) wrote…</i></summary><blockquote>

I think bool return is ok for context like this
</blockquote></details>

Done.
___
*[libraries/chain/hardfork.d/20200320.hf, line 4 at r1](https://reviewable.io/reviews/cocos-bcx/cocos-mainnet/76#-M2qs2o7zm-Dm5xroLPu-r1-4:-M2rLRMAAuCAKceVKQmD:b-896fix) ([raw file](https://github.com/cocos-bcx/cocos-mainnet/blob/8335277d80aa74ffbf4755a4b8f1ee9047d7cc5f/libraries/chain/hardfork.d/20200320.hf#L4)):*
<details><summary><i>Previously, wanliqun (wanliqun) wrote…</i></summary><blockquote>

I observed this macro had also been defined in libraries/chain/include/graphene/chain/hardfork.hpp. Why do we need to define this in two different places?
</blockquote></details>

Done.
___
*[libraries/chain/include/graphene/chain/hardfork.hpp, line 27 at r1](https://reviewable.io/reviews/cocos-bcx/cocos-mainnet/76#-M2qs2o7zm-Dm5xroLPw-r1-27:-M2rLV2CElAi_8KQsydo:b-896fix) ([raw file](https://github.com/cocos-bcx/cocos-mainnet/blob/8335277d80aa74ffbf4755a4b8f1ee9047d7cc5f/libraries/chain/include/graphene/chain/hardfork.hpp#L27)):*
<details><summary><i>Previously, wanliqun (wanliqun) wrote…</i></summary><blockquote>

Why use such name as UNAUTO_GAS?  Would be much better to use a more explicit name.
</blockquote></details>

Done.

**Review by wanliqun — APPROVED** [review]

**Pull request merged**